### PR TITLE
[action] add `address` and `hash` field in SealedEnvelope

### LINKED ISSUE #2989: Add `address` and `hash` field in SealedEnvelope struct
**What would you like to be added**:

**Why is this needed**:
Doing `selp.SrcPubkey().Address()` and `selp.Hash()` calculation on the fly is time-consuminig. It can be calculated once when loadProto, and be reused later.

### PR DESCRIPTION
**Pull request by saitofun**
# Description
    - perf: [action] improve SealedEnvelope accessing `hash`, `hashErr` and `srcAddress`
    - test: [action] improve TestSealedEnvelope_Basic TestSealedEnvelope_Proto and TestActionDeserializer
    - feat: [action] add `ErrInvalidSigLen` and `ErrUnknownEncType` error func for maintain test and error better

Fixes #2989 

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement
- [x] New feature (non-breaking change which adds functionality)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/sealedenvelope.go [hunk 1]
```diff
@@ -48,6 +63,15 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
 // Hash returns the hash value of SealedEnvelope.
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
+	if sealed.hash == nil {
+		hv, err := sealed.calcHash()
+		sealed.hash = &Hash256Result{hv, err}
+	}
+	return sealed.hash.Value()
```
Comment (dustinxie) [inline action/sealedenvelope.go:70]: `return sealed.hash.Hash256, sealed.hash.error`

File: action/sealedenvelope.go [hunk 2]
```diff
@@ -22,6 +23,20 @@ type SealedEnvelope struct {
...
+	address      address.Address
+	hash         *Hash256Result
+}
+
+type Hash256Result struct {
+	hash.Hash256
+	error
+}
+
+func (h *Hash256Result) Value() (hash.Hash256, error) {
```
Comment (dustinxie) [inline action/sealedenvelope.go:35]: no need for this func, see comments below

File: action/sealedenvelope.go [hunk 3]
```diff
@@ -22,6 +23,20 @@ type SealedEnvelope struct {
 	evmNetworkID uint32
 	srcPubkey    crypto.PublicKey
 	signature    []byte
+	address      address.Address
+	hash         *Hash256Result
+}
+
+type Hash256Result struct {
```
Comment (dustinxie) [inline action/sealedenvelope.go:30]: `hash256Result`, this is internal data struct

File: action/sealedenvelope_test.go [hunk 4]
```diff
@@ -35,7 +35,7 @@ func TestSealedEnvelope_Basic(t *testing.T) {
 		{0, "322884fb04663019be6fb461d9453827487eafdd57b4de3bd89a7d77c9bf8395"},
 		{1, "80af7840d73772d3022d8bdc46278fb755352e5e9d5f2a1f12ee7ec4f1ea98e9"},
 	} {
-		se, err := createSealedEnvelope(v.id)
+		se, err := createSealedEnvelopeWithSignature(v.id, _signByte)
```
Comment (dustinxie) [inline action/sealedenvelope_test.go:38]: should not need this? see comments below

File: action/sealedenvelope_test.go [hunk 5]
```diff
@@ -198,8 +199,42 @@ func createSealedEnvelope(chainID uint32) (SealedEnvelope, error) {
...
+		return se, err
+	}
+
 	se.Envelope = evlp
 	se.srcPubkey = cPubKey
-	se.signature = _signByte
+	se.signature = _validSig
+	if err = se.LoadProto(se.Proto()); err != nil {
+		return se, err
+	}
```
Comment (dustinxie) [inline action/sealedenvelope_test.go:211]: we don't need this? see comments below

File: action/sealedenvelope_test.go [hunk 6]
```diff
@@ -198,8 +199,42 @@ func createSealedEnvelope(chainID uint32) (SealedEnvelope, error) {
...
 	return se, err
 }
+
+func createSealedEnvelopeWithSignature(chainID uint32, sig []byte) (SealedEnvelope, error) {
```
Comment (dustinxie) [inline action/sealedenvelope_test.go:215]: we don't need to add this func? if there's a need (for test) about signature, just directly set the signature (as done in tests now)

File: action/sealedenvelope.go [hunk 7]
```diff
@@ -69,6 +93,13 @@ func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
 // SrcPubkey returns the source public key
 func (sealed *SealedEnvelope) SrcPubkey() crypto.PublicKey { return sealed.srcPubkey }
 
+func (sealed *SealedEnvelope) Address() address.Address {
```
Comment (dustinxie) [inline action/sealedenvelope.go:96]: need a comment like // Address() to satisfy `go lint`
you run `make lint` local to make sure no lint error

File: action/sealedenvelope.go (hunk 3 above)
Comment (Liuhaai) [reply action/sealedenvelope.go:30]: Remove this struct

File: action/action_deserializer_test.go [hunk 8]
```diff
@@ -10,34 +10,50 @@ import (
...
+				ActionToSealedEnvelope(se.Proto())
+			NewWithT(t).Expect(validated).NotTo(Equal(se))
+
+			NewWithT(t).Expect(se.hash).To(Equal(hash.ZeroHash256))
+			NewWithT(t).Expect(se.hashErr).To(BeNil())
+			_, _ = se.Hash()
+			_ = se.Address()
+			// after Hash() and Address() was invoked, the two value are equal
+			NewWithT(t).Expect(validated).To(Equal(se))
+		})
```
Comment (dustinxie) [inline action/action_deserializer_test.go:56]: revert these, we don't want to introduce a new testing style, pls only do minimum necessary changes to make existing test pass

File: action/const.go [hunk 9]
```diff
@@ -32,6 +35,15 @@ var (
...
 
+var (
+	ErrInvalidSigLen = func(length int) error {
+		return errors.Errorf("invalid signature length = %d, expecting 65", length)
+	}
+	ErrUnknownEncType = func(enc iotextypes.Encoding) error {
+		return errors.Errorf("unknown encoding type %d", enc)
+	}
+)
+
```
Comment (dustinxie) [inline action/const.go:46]: not very necessary, seems an overkill

File: action/sealedenvelope.go [hunk 10]
```diff
@@ -63,41 +53,41 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
...
-		hv, err := sealed.calcHash()
-		sealed.hash = &Hash256Result{hv, err}
+	if sealed.hash != hash.ZeroHash256 || sealed.hashErr != nil {
+		return sealed.hash, sealed.hashErr
 	}
-	return sealed.hash.Value()
-}
-
-// calcHash returns the hash value of SealedEnvelope
-func (sealed *SealedEnvelope) calcHash() (hash.Hash256, error) {
```
Comment (dustinxie) [inline action/sealedenvelope.go]: keep this, it is clear to have an internal func (which might be used)

File: action/sealedenvelope.go (hunk 7 above)
Comment (dustinxie) [reply action/sealedenvelope.go:86]: change to `SenderAddress`, it's a better/clear name

File: action/sealedenvelope.go [hunk 11]
```diff
@@ -151,43 +142,17 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, withChainID bo
...
-	sealed.signature = make([]byte, sigSize)
-	copy(sealed.signature, pbAct.GetSignature())
-	sealed.encoding = encoding
-	hashValue, err := sealed.calcHash()
-	sealed.hash = &Hash256Result{hashValue, err}
-	sealed.address = srcPub.Address()
-	return nil
+	sealed.encoding = pbAct.GetEncoding()
+	_, err = sealed.Hash()
+	return err
```
Comment (dustinxie) [inline action/sealedenvelope.go:155]: revert, like we discussed, there's no need to change `loadProto` at all

File: action/sealedenvelope_test.go [hunk 12]
```diff
@@ -129,90 +130,110 @@ func TestSealedEnvelope_Actions(t *testing.T) {
...
+			// after Hash and Address invoked, validated should be equal to se
+			NewWithT(t).Expect(se.srcAddress).To(Equal(validated.srcAddress))
+			NewWithT(t).Expect(se.hash).To(Equal(validated.hash))
+			NewWithT(t).Expect(se).To(Equal(validated))
+
+			tsf1, ok1 := se.Envelope.Action().(*Transfer)
+			tsf2, ok2 := validated.Envelope.Action().(*Transfer)
+			NewWithT(t).Expect(tsf1).To(Equal(tsf2))
+			NewWithT(t).Expect(ok1 && ok2).To(BeTrue())
+		})
```
Comment (dustinxie) [inline action/sealedenvelope_test.go:232]: no need to change these test files

File: action/sealedenvelope.go [hunk 13]
```diff
@@ -63,41 +53,41 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
 // Hash returns the hash value of SealedEnvelope.
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
-	if sealed.hash == nil {
-		hv, err := sealed.calcHash()
-		sealed.hash = &Hash256Result{hv, err}
+	if sealed.hash != hash.ZeroHash256 || sealed.hashErr != nil {
+		return sealed.hash, sealed.hashErr
```
Comment (dustinxie) [inline action/sealedenvelope.go:57]: just like the 1st commit
```
if sealed.hash == hash.ZeroHash256 && sealed.hashErr == nil {
    sealed.hash, sealed.hashErr = sealed.calcHash()
}
return sealed.calcHash()
```

File: action/action_deserializer_test.go (hunk 8 above)
Comment (dustinxie) [reply action/action_deserializer_test.go:56]: as we discussed, please revert all unnecessary changes, only need these 3 lines to make the test pass
```
_, err = se.Hash()
r.NoError(err)
se.SenderAddress()
// after Hash() and Address() was invoked, the two value are equal
```

File: action/const.go (hunk 9 above)
Comment (dustinxie) [reply action/const.go:46]: as we discussed, we don't really need this

File: action/sealedenvelope.go [hunk 14]
```diff
@@ -48,27 +52,43 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
...
+			break
 		}
 		tx, err := act.ToEthTx()
 		if err != nil {
-			return hash.ZeroHash256, err
+			sealed.hash, sealed.hashErr = hash.ZeroHash256, err
+			break
 		}
-		return rlpSignedHash(tx, sealed.evmNetworkID, sealed.Signature())
+		sealed.evmNetworkID = config.EVMNetworkID()
```
Comment (dustinxie) [inline action/sealedenvelope.go:70]: why do this? the correct place to set `evmNetworkID` is in `LoadProto()`

File: action/sealedenvelope.go [hunk 15]
```diff
@@ -48,27 +52,43 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
...
+		sealed.hash, sealed.hashErr =
+			rlpSignedHash(tx, sealed.evmNetworkID, sealed.Signature())
 	case iotextypes.Encoding_IOTEX_PROTOBUF:
-		return hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto()))), nil
+		sealed.hash = hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto())))
 	default:
-		return hash.ZeroHash256, errors.Errorf("unknown encoding type %v", sealed.encoding)
+		sealed.hashErr = ErrUnknownEncType(sealed.encoding)
 	}
+	return sealed.hash, sealed.hashErr
```
Comment (dustinxie) [inline action/sealedenvelope.go:78]: as we discussed, please revert unnecessary changes, and make this an internal func `calcHash()`

File: action/sealedenvelope.go [hunk 16]
```diff
@@ -153,6 +173,8 @@ func (sealed *SealedEnvelope) loadProto(pbAct *iotextypes.Action, withChainID bo
 	sealed.signature = make([]byte, sigSize)
 	copy(sealed.signature, pbAct.GetSignature())
 	sealed.encoding = encoding
+	sealed.Hash()
+	sealed.SenderAddress()
```
Comment (dustinxie) [inline action/sealedenvelope.go:177]: as we discussed, not necessary

File: action/sealedenvelope_test.go [hunk 17]
```diff
@@ -144,16 +144,29 @@ func TestSealedEnvelope_Proto(t *testing.T) {
...
+		},
 	} {
-		se.encoding = v.encoding
-		se.signature = v.sig
-		req.Contains(se2.LoadProto(se.Proto()).Error(), v.err)
+		t.Run(v.name, func(t *testing.T) {
+			se.encoding = v.encoding
+			se.signature = v.sig
+			req.Equal(se2.LoadProto(se.Proto()).Error(), v.err.Error())
+		})
```
Comment (dustinxie) [inline action/sealedenvelope_test.go:169]: as we discussed, please revert unnecessary changes, do minimum changes to make the test pass

File: action/sealedenvelope.go [hunk 18]
```diff
@@ -48,27 +52,46 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
...
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
+	if sealed.hash != hash.ZeroHash256 || sealed.hashErr != nil {
+		return sealed.hash, sealed.hashErr
+	}
+	sealed.calcHash()
+	return sealed.hash, sealed.hashErr
+}
+
+func (sealed *SealedEnvelope) calcHash() {
```
Comment (dustinxie) [inline action/sealedenvelope.go:62]: one final comment:
```
func (sealed *SealedEnvelope) calcHash() (hash.Hash256, error)
```
then don't need to change this func at all, in the above, just
```
sealed.hash, sealed.hashErr = calcHash()
```

**Review by dustinxie — APPROVED** [review]

File: action/const.go [hunk 19]
```diff
@@ -6,7 +6,9 @@
 
 package action
 
-import "github.com/pkg/errors"
+import (
```
Comment (Liuhaai) [inline action/const.go:9]: revert

File: action/sealedenvelope.go [hunk 20]
```diff
@@ -5,12 +5,13 @@ import (
 
 	"github.com/iotexproject/go-pkgs/crypto"
 	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
```
Comment (Liuhaai) [inline action/sealedenvelope.go:8]: revert

File: action/sealedenvelope.go [hunk 21]
```diff
@@ -22,6 +23,9 @@ type SealedEnvelope struct {
 	evmNetworkID uint32
 	srcPubkey    crypto.PublicKey
 	signature    []byte
+	srcAddress   address.Address
+	hash         hash.Hash256
+	hashErr      error
```
Comment (Liuhaai) [inline action/sealedenvelope.go:28]: remove

File: action/const.go (hunk 19 above)
Comment (saitofun) [reply action/const.go]: reverted

File: action/sealedenvelope.go (hunk 20 above)
Comment (saitofun) [reply action/sealedenvelope.go:8]: reverted

File: actpool/actpool_test.go [hunk 22]
```diff
@@ -902,7 +902,12 @@ func TestActPool_GetUnconfirmedActs(t *testing.T) {
 	require.Equal([]action.SealedEnvelope(nil), acts)
 
 	acts = ap.GetUnconfirmedActs(_addr1)
-	require.Equal([]action.SealedEnvelope{tsf1, tsf3, tsf4, tsf5}, acts)
+	validated := []action.SealedEnvelope{tsf1, tsf3, tsf4, tsf5}
```
Comment (Liuhaai) [inline actpool/actpool_test.go:905]: revert

File: actpool/actpool_test.go [hunk 23]
```diff
@@ -1089,7 +1094,12 @@ func TestActPool_SpeedUpAction(t *testing.T) {
 		appliedActionList = append(appliedActionList, bestAction)
 	}
 	// tsf1 is replaced by tsf3 with higher gas price
-	require.Equal(appliedActionList, []action.SealedEnvelope{tsf3, tsf2})
+	validated := []action.SealedEnvelope{tsf3, tsf2}
```
Comment (Liuhaai) [inline actpool/actpool_test.go:1097]: revert

File: blockchain/blockdao/blockdao_test.go [hunk 24]
```diff
@@ -203,9 +203,17 @@ func TestBlockDAO(t *testing.T) {
 			require.Equal(tipBlk.Height(), height)
 			blk, err := dao.GetBlock(hash)
 			require.NoError(err)
+			for idx := range blk.Actions {
```
Comment (Liuhaai) [inline blockchain/blockdao/blockdao_test.go:206]: revert

File: action/action_deserializer_test.go [hunk 25]
```diff
@@ -34,8 +35,12 @@ func TestActionDeserializer(t *testing.T) {
 		r.Equal(_signByte, se.Signature())
 		r.Zero(se.Encoding())
 
+		// use valid signature and reset se.Hash
 		se.signature = _validSig
+		se.hash = hash.ZeroHash256
+		se.Hash()
 		se1, err := (&Deserializer{}).ActionToSealedEnvelope(se.Proto())
+		se1.Hash()
```
Comment (CoderZhi) [inline action/action_deserializer_test.go:43]: same here

**Review by Liuhaai — APPROVED** [review]

File: actpool/actpool_test.go [hunk 26]
```diff
@@ -902,7 +902,12 @@ func TestActPool_GetUnconfirmedActs(t *testing.T) {
 	require.Equal([]action.SealedEnvelope(nil), acts)
 
 	acts = ap.GetUnconfirmedActs(_addr1)
-	require.Equal([]action.SealedEnvelope{tsf1, tsf3, tsf4, tsf5}, acts)
+	validated := []action.SealedEnvelope{tsf1, tsf3, tsf4, tsf5}
+	for i := range validated {
+		_, err = validated[i].Hash()
+		require.NoError(err)
+	}
+	require.Equal(validated, acts)
```
Comment (Liuhaai) [inline actpool/actpool_test.go:910]: asserting hash
